### ISSUE TITLE
SQL pool frequently marked as overloaded

### ISSUE THREAD

**Issue by wichert** [processone/ejabberd#3016]
## Environment
SQL connection pool is overloaded, discarding stale requests

- ejabberd version: 19.08
- Erlang version: `Erlang (SMP,ASYNC_THREADS,HIPE) (BEAM) emulator version 10.2.1`
- OS: ?
- Installed from: `ejabberd/ecs:19.08` docker image

## Configuration (only if needed): grep -Ev '^$|^\s*#' ejabberd.yml

```yaml
hosts:
  - "company.com"
loglevel: 4
log_rotate_size: 10485760
log_rotate_date: ""
log_rotate_count: 3
log_rate_limit: 100
listen:
  -
    port: 5222
    ip: "::"
    module: ejabberd_c2s
    max_stanza_size: 65536
    shaper: c2s_shaper
    access: c2s
  -
    port: 5223
    module: ejabberd_c2s
    starttls: true
    starttls_required: true
    dhfile: /home/ejabberd/dhparam/dhparam.pem
    protocol_options:
      - "no_sslv3"
      - "no_tlsv1"
      - "no_tlsv1_1"
      - "cipher_server_preference"
      - "no_compression"
    max_stanza_size: 65536
    shaper: c2s_shaper
    access: c2s
  -
    port: 5269
    ip: "::"
    module: ejabberd_s2s_in
    max_stanza_size: 524288
  -
    port: 5280
    ip: "::"
    module: ejabberd_http
    request_handlers:
      "/": ejabberd_http_ws
  -
    port: 5347
    module: ejabberd_service
    access: all
    shaper: fast
    ip: "::"
    hosts: {}
  -
    port: 5281
    ip: "::"
    module: ejabberd_http
    request_handlers:
      "/admin": ejabberd_web_admin
      "/api": mod_http_api
sql_type: pgsql
sql_server: "postgresql"
sql_database: "XXX"
sql_username: "XXX"
sql_password: "XXX"
sql_port: 5432
default_db: sql
acl:
  admin:
    user:
      - "admin": "company.com"
  local:
    user_regexp: ""
  loopback:
    ip:
      - "127.0.0.0/8"
      - "::1/128"
      - ::FFFF:127.0.0.1/128
access_rules:
  local:
    allow: local
  c2s:
    deny: blocked
    allow: all
  announce:
    allow: admin
  configure:
    allow: admin
  muc_admin:
    allow: admin
  muc_create:
    allow: local
  muc:
    allow: all
  pubsub_createnode:
    allow: admin
api_permissions:
  "console commands":
    from:
      - ejabberd_ctl
    who: all
    what: "*"
  "admin access":
    who:
      access:
        allow:
          acl: loopback
          acl: admin
      oauth:
        scope: "ejabberd:admin"
        access:
          allow:
            acl: loopback
            acl: admin
    what:
      - "*"
      - "!stop"
      - "!start"
shaper:
  normal: 5000
  fast: 50000
shaper_rules:
  max_user_sessions: 10
  max_user_offline_messages:
    5000: admin
    100: all
  c2s_shaper:
    none: admin
    normal: all
  s2s_shaper: fast
max_fsm_queue: 10000
acme:
  contact: "mailto:me@example.com"
  ca_url: "https://acme-v01.api.letsencrypt.org"
modules:
  mod_adhoc: {}
  mod_announce:
    access: announce
  mod_blocking: {}
  mod_caps: {}
  mod_carboncopy: {}
  mod_client_state:
    queue_presence: false
  mod_configure: {}
  mod_disco: {}
  mod_http_api: {}
  mod_last:
    db_type: sql
  mod_muc:
    access: muc
    access_create: muc_create
    access_persistent: muc_create
    access_admin: muc_admin
  mod_offline:
    db_type: sql
    access_max_user_messages: max_user_offline_messages
  mod_ping: {}
  mod_privacy:
    db_type: sql
  mod_private:
    db_type: sql
  mod_pubsub:
    access_createnode: pubsub_createnode
    db_type: sql
    last_item_cache: false
    plugins:
      - "flat"
      - "pep" # pep requires mod_caps
    max_items_node: 1000
  mod_push: {}
  mod_roster:
    db_type: sql
    versioning: true
    store_current_id: false
    use_cache: false
  mod_shared_roster: {}
  mod_stats: {}
  mod_time: {}
  mod_vcard:
    search: false
    db_type: sql
  mod_version:
    show_os: false
allow_contrib_modules: true

```

## Errors from error.log/crash.log

```
13:58:12.639 [error] SQL connection pool is overloaded, discarding stale requests
```

## Bug description

We are testing switching to Ejabberd 19.08 (upgrading from 16.06). On a server with only a few people (2 - 60) with very little traffic, I still frequently see `SQL connection pool is overloaded` errors appear. When that happens ejabberd stops responding to everything, until it hits some internal timeout and suddenly starts working again.

When this happens PostgreSQL does not show any connections actually being in use:

```
11393 ?        Ss     0:28 postgres: ejabberd crypho 192.168.61.93(40497) idle
11394 ?        Ss     0:15 postgres: ejabberd crypho 192.168.61.93(44817) idle
11395 ?        Ss     0:02 postgres: ejabberd crypho 192.168.61.93(38277) idle
11396 ?        Ss     0:15 postgres: ejabberd crypho 192.168.61.93(52091) idle
11397 ?        Ss     0:15 postgres: ejabberd crypho 192.168.61.93(36909) idle
11398 ?        Ss     0:15 postgres: ejabberd crypho 192.168.61.93(41171) idle
11399 ?        Ss     0:09 postgres: ejabberd crypho 192.168.61.93(48049) idle
11400 ?        Ss     0:01 postgres: ejabberd crypho 192.168.61.93(57539) idle
11401 ?        Ss     0:15 postgres: ejabberd crypho 192.168.61.93(58679) idle
11402 ?        Ss     0:15 postgres: ejabberd crypho 192.168.61.93(34879) idle
```

**Comment by zinid** [comment]
Do you see other error messages in error.log?

**Comment by wichert** [comment]
There are no other errors. The first time this happened for one run looks like this:

```
09:31:18.507 [info] (<0.726.0>) Accepted connection [::ffff:192.168.61.107]:38594 -> [::ffff:192.168.61.93]:5280
09:31:18.706 [info] (websocket|<0.727.0>) Accepted c2s PLAIN authentication for username@domain by sql backend from ::ffff:192.168.61.107
09:31:18.839 [info] (websocket|<0.727.0>) Opened c2s session for username@domain/web-b381f545-59b3-41fb-8833-c319daeb8e45
09:31:48.319 [error] SQL connection pool is overloaded, discarding stale requests
```

Notice there are 30 seconds without any log messages before the overload happened.

**Comment by zinid** [comment]
Since the load is not high, I recommend setting ejabberd to debug mode (`loglevel: 5`), enabling log query in PostgreSQL and monitor this stuff closely. If you don't find anything suspicious, post the logs here. If we don't find anything we can trace SQL pool directly from inside the Erlang emulator, but this is tricky (and that's why I suggest doing this as a last resort).

**Comment by wichert** [comment]
Will do.

**Comment by zinid** [comment]
Any update on this? Please update the info if you want the fix to go into 19.10.

**Comment by wichert** [comment]
Due to #3025 we had to revert back to an old ejabberd version, so I'm afraid I can't test this any further. Feel free to close it if you want; if I run into it again once we manage to a newer ejabberd version I can create a new issue.
